<head>
 <meta http-equiv="Content-Type" content="text/html ">
 <title>'[PATCH 4/6] blk-mq: avoid to map CPU into stale hw queue' - MARC</title>
 <link rel="stylesheet" type="text/css" href="/style.css">
</head>

<body bgcolor="#ffffff" text="#000000" link="#0000C0" vlink="#800080">
<pre><b>[<a href="?l=linux-block&m=151307662318723&w=2">prev in list</a>] [<a href="?l=linux-block&m=151307665618757&w=2">next in list</a>] [<font color="#c0c0c0">prev in thread</font>] [<a href="?l=linux-block&m=151308796622799&w=2">next in thread</a>] </b>
<b><font size=+1>
List:       <a href="?l=linux-block&r=1&w=2">linux-block</a>
Subject:    <a href="?t=151307673000008&r=1&w=2">[PATCH 4/6] blk-mq: avoid to map CPU into stale hw queue</a>
From:       <a href="?a=149155772200003&r=1&w=2">Ming Lei &lt;ming.lei () redhat ! com&gt;</a>
Date:       <a href="?l=linux-block&r=1&w=2&b=201712">2017-12-12 11:02:30</a>
Message-ID: <a href="?i=20171212110232.12495-5-ming.lei%20()%20redhat%20!%20com">20171212110232.12495-5-ming.lei () redhat ! com</a></font>
[<a href="?l=linux-block&m=151307663318727&q=raw">Download message RAW</a>]</b>

blk_mq_pci_map_queues() may not map one CPU into any hw queue, but its
previous map isn't cleared yet, and may point to one stale hw queue
index.

This patch fixes the following issue by clearing the mapping table before
setting it up in blk_mq_pci_map_queues().

This patches fixes this following issue reported by Zhang Yi:

[  101.202734] BUG: unable to handle kernel NULL pointer dereference at \
0000000094d3013f [  101.211487] IP: blk_mq_map_swqueue+0xbc/0x200
[  101.216346] PGD 0 P4D 0
[  101.219171] Oops: 0000 [#1] SMP
[  101.222674] Modules linked in: sunrpc ipmi_ssif vfat fat intel_rapl sb_edac \
x86_pkg_temp_thermal intel_powerclamp coretemp kvm_intel kvm irqbypass \
crct10dif_pclmul crc32_pclmul ghash_clmulni_intel intel_cstate intel_uncore mxm_wmi \
intel_rapl_perf iTCO_wdt ipmi_si ipmi_devintf pcspkr iTCO_vendor_support sg dcdbas \
ipmi_msghandler wmi mei_me lpc_ich shpchp mei acpi_power_meter dm_multipath ip_tables \
xfs libcrc32c sd_mod mgag200 i2c_algo_bit drm_kms_helper syscopyarea sysfillrect \
sysimgblt fb_sys_fops ttm drm ahci libahci crc32c_intel libata tg3 nvme nvme_core \
megaraid_sas ptp i2c_core pps_core dm_mirror dm_region_hash dm_log dm_mod [  \
101.284881] CPU: 0 PID: 504 Comm: kworker/u25:5 Not tainted 4.15.0-rc2 #1 [  \
101.292455] Hardware name: Dell Inc. PowerEdge R730xd/072T6D, BIOS 2.5.5 08/16/2017 [ \
101.301001] Workqueue: nvme-wq nvme_reset_work [nvme] [  101.306636] task: \
00000000f2c53190 task.stack: 000000002da874f9 [  101.313241] RIP: \
0010:blk_mq_map_swqueue+0xbc/0x200 [  101.318681] RSP: 0018:ffffc9000234fd70 EFLAGS: \
00010282 [  101.324511] RAX: ffff88047ffc9480 RBX: ffff88047e130850 RCX: \
0000000000000000 [  101.332471] RDX: ffffe8ffffd40580 RSI: ffff88047e509b40 RDI: \
ffff88046f37a008 [  101.340432] RBP: 000000000000000b R08: ffff88046f37a008 R09: \
0000000011f94280 [  101.348392] R10: ffff88047ffd4d00 R11: 0000000000000000 R12: \
ffff88046f37a008 [  101.356353] R13: ffff88047e130f38 R14: 000000000000000b R15: \
ffff88046f37a558 [  101.364314] FS:  0000000000000000(0000) GS:ffff880277c00000(0000) \
knlGS:0000000000000000 [  101.373342] CS:  0010 DS: 0000 ES: 0000 CR0: \
0000000080050033 [  101.379753] CR2: 0000000000000098 CR3: 000000047f409004 CR4: \
00000000001606f0 [  101.387714] Call Trace:
[  101.390445]  blk_mq_update_nr_hw_queues+0xbf/0x130
[  101.395791]  nvme_reset_work+0x6f4/0xc06 [nvme]
[  101.400848]  ? pick_next_task_fair+0x290/0x5f0
[  101.405807]  ? __switch_to+0x1f5/0x430
[  101.409988]  ? put_prev_entity+0x2f/0xd0
[  101.414365]  process_one_work+0x141/0x340
[  101.418836]  worker_thread+0x47/0x3e0
[  101.422921]  kthread+0xf5/0x130
[  101.426424]  ? rescuer_thread+0x380/0x380
[  101.430896]  ? kthread_associate_blkcg+0x90/0x90
[  101.436048]  ret_from_fork+0x1f/0x30
[  101.440034] Code: 48 83 3c ca 00 0f 84 2b 01 00 00 48 63 cd 48 8b 93 10 01 00 00 \
8b 0c 88 48 8b 83 20 01 00 00 4a 03 14 f5 60 04 af 81 48 8b 0c c8 &lt;48&gt; 8b 81 98 00 00 \
00 f0 4c 0f ab 30 8b 81 f8 00 00 00 89 42 44 [  101.461116] RIP: \
blk_mq_map_swqueue+0xbc/0x200 RSP: ffffc9000234fd70 [  101.468205] CR2: \
0000000000000098 [  101.471907] ---[ end trace 5fe710f98228a3ca ]---
[  101.482489] Kernel panic - not syncing: Fatal exception
[  101.488505] Kernel Offset: disabled
[  101.497752] ---[ end Kernel panic - not syncing: Fatal exception

Reported-by: Yi Zhang &lt;yi.zhang@redhat.com&gt;
Signed-off-by: Ming Lei &lt;ming.lei@redhat.com&gt;
---
 block/blk-mq-pci.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/block/blk-mq-pci.c b/block/blk-mq-pci.c
index 76944e3271bf..b62212cb8060 100644
--- a/block/blk-mq-pci.c
+++ b/block/blk-mq-pci.c
@@ -33,6 +33,14 @@ int blk_mq_pci_map_queues(struct blk_mq_tag_set *set, struct \
pci_dev *pdev)  const struct cpumask *mask;
 	unsigned int queue, cpu;
 
+	/*
+	 * When nr_hw_queue is resized, we have to reset the map table
+	 * for avoiding stale mapping since one CPU may not be mapped
+	 * to any hw queue.
+	 */
+	for_each_possible_cpu(cpu)
+		set-&gt;mq_map[cpu] = 0;
+
 	for (queue = 0; queue &lt; set-&gt;nr_hw_queues; queue++) {
 		mask = pci_irq_get_affinity(pdev, queue);
 		if (!mask)
-- 
2.9.5


<b>[<a href="?l=linux-block&m=151307662318723&w=2">prev in list</a>] [<a href="?l=linux-block&m=151307665618757&w=2">next in list</a>] [<font color="#c0c0c0">prev in thread</font>] [<a href="?l=linux-block&m=151308796622799&w=2">next in thread</a>] </b>
</pre>
  </pre><br><center>
    <a href="?q=configure">Configure</a> | 

    <a href="?q=about">About</a> |
    <a href="?q=news">News</a> |
    <a href="mailto:webguy@marc.info?subject=Add%20a%20list%20to%20MARC">Add&nbsp;a&nbsp;list</a> |
    Sponsored&nbsp;by&nbsp;<a href="http://www.korelogic.com/">KoreLogic</a>
</center>
</body>
</html>
